src/Database/Franz/Server.hs

```diff
@@ -0,0 +1,186 @@
+{-# LANGUAGE DeriveGeneric, LambdaCase, OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE BangPatterns #-}
+module Database.Franz.Server
+  ( Settings(..)
+  , startServer
+  , defaultPort
+  ) where
+
+import Control.Concurrent
+import Control.Exception
+import Control.Monad
+import Control.Retry
+import Control.Concurrent.STM
+import Database.Franz.Internal
+import Database.Franz.Protocol
+import Database.Franz.Reader
+import Data.ConcurrentResourceMap
+import Data.Serialize
+import qualified Data.IntMap.Strict as IM
+import Data.IORef
+import qualified Data.ByteString.Char8 as B
+import qualified Data.HashMap.Strict as HM
+import Data.Tuple (swap)
+import qualified Network.Socket.SendFile.Handle as SF
+import qualified Network.Socket.ByteString as SB
+import qualified Network.Socket as S
+import System.Directory
+import System.FilePath
+import System.IO
+import System.Process (ProcessHandle, spawnProcess, cleanupProcess,
+  waitForProcess, getProcessExitCode, getPid)
+
+respond :: FranzReader
+  -> IORef (IM.IntMap ThreadId)
+  -> FilePath
+  -> IORef B.ByteString
+  -> MVar S.Socket -> IO ()
+respond env refThreads path buf vConn = do
+  recvConn <- readMVar vConn
+  runGetRecv buf recvConn get >>= \case
+    Right (RawRequest reqId req) -> do
+      (stream, query) <- handleQuery env path req
+      touchActivity stream
+      atomically (fmap Right query `catchSTM` (pure . Left)) >>= \case
+        Left e -> sendHeader $ ResponseError reqId e
+        Right (ready, offsets)
+          | ready -> send (ResponseInstant reqId) stream offsets
+          | otherwise -> do
+            m <- readIORef refThreads
+            if IM.member reqId m
+              then sendHeader $ ResponseError reqId $ MalformedRequest "duplicate request ID"
+              else do
+                sendHeader $ ResponseWait reqId
+                -- Fork a thread to send a delayed response
+                let finish _ = do
+                      _ <- popThread reqId
+                      removeActivity stream
+
+                tid <- flip forkFinally finish $ do
+                  addActivity stream
```

Comment:
I'd just like to point out that this is subject to not fire if thread dies instantly, so we could easily end up running more `removeActivity` than `addActivity`...